apis/python/src/tiledbsoma/collection.py

```diff
@@ -168,12 +176,70 @@ def add_new_collection(
         self._close_stack.enter_context(new_group)
         return new_group
 
-    def _not_implemented(self, *args: Any, **kwargs: Any) -> NoReturn:
-        raise NotImplementedError()
+    def add_new_dataframe(
+        self,
+        key: str,
+        *,
+        uri: Optional[str] = None,
+        schema: pa.Schema,
+        index_column_names: Sequence[str] = (SOMA_JOINID,),
+        platform_config: Optional[options.PlatformConfig] = None,
+    ) -> DataFrame:
+        if key in self:
+            raise KeyError(f"{key!r} already exists in {type(self)}")
+        absolute_uri, was_relative = self._new_absolute_uri(key=key, uri=uri)
+        self._check_allows_child(key, DataFrame)
+        new_df = DataFrame.create(
+            absolute_uri,
+            index_column_names=index_column_names,
+            schema=schema,
+            platform_config=platform_config,
+            context=self.context,
+        )
+        # A dataframe might not be the declared type of this collection,
+        # but we can't really handle that within the type system.
```

Comment:
curious, what typing check are we unable to validate in `_check_allows_child()`, above?

Reply:
Let's say you have a `Collection[Measurement]`. You can do `meas_coll.add_new_df(...)` despite the fact that (in theory) the only thing you're *supposed* to be able to add is a `Measurement`. The type system won't warn you that you're about to do something dumb in that case (and we don't have a mechanism to do so at runtime, in part because doing so would add additional complexity).

Reply:
Uhh ... I didn't touch this file. Just ran `python -m pre_commit run -a -v` ...